src/time-wait-sync/time-wait-sync.c

```diff
@@ -0,0 +1,176 @@
+        if (r < 0) {
+                log_error_errno(r, "Failed to set timerfd conditions: %m");
+                goto finish;
+        }
+
+        r = adjtimex(&tx);
+        if (r < 0) {
+                log_error_errno(r, "Failed to read adjtimex state: %m");
+                goto finish;
+        }
+        sp->adjtime_state = r;
+
+        if (tx.status & STA_NANO) {
+                tx.time.tv_usec /= 1000;
+        }
+        t = timeval_load(&tx.time);
+        ts = format_timestamp_us_utc(buf, sizeof(buf), t);
+        if (!ts) {
+                strcpy(buf, "unrepresentable");
+        }
+        fprintf(stderr, "adjtime state %d status %x time %s\n", sp->adjtime_state, tx.status, ts);
+
+        if (sp->adjtime_state == TIME_ERROR) {
+                /* Not synchronized.  Do a one-shot wait on the descriptor and inform the caller we need to keep
+                 * running. */
+                r = sd_event_add_io(event, &sp->event_source, sp->fd,
+                                    EPOLLIN, io_handler, sp);
+                if (r < 0) {
+                        log_error_errno(r, "Failed to create time change monitor source: %m");
+                        goto finish;
+                }
+                r = 1;
+        } else {
+                /* Synchronized; we can exit. */
+                (void) sd_event_exit(event, 0);
+                r = 0;
+        }
+
+ finish:
+        if (r < 0)
+                (void) sd_event_exit(event, r);
+        return r;
+}
+
+int main(int argc,
+         char * argv[]) {
+        int r;
+        sd_event *event;
+        ClockState state = {
+                .fd = -1,
+        };
+
+        assert_se(sigprocmask_many(SIG_BLOCK, NULL, SIGTERM, SIGINT, -1) >= 0);
+
+        r = sd_event_default(&event);
+        if (r < 0) {
+                log_error_errno(r, "Failed to allocate event loop: %m");
+                goto finish;
+        }
+
+        r = sd_event_add_signal(event, NULL, SIGTERM, NULL, NULL);
+        if (r < 0) {
+                log_error_errno(r, "Failed to create sigterm event source: %m");
+                goto finish;
+        }
+
+        r = sd_event_add_signal(event, NULL, SIGINT, NULL, NULL);
+        if (r < 0) {
+                log_error_errno(r, "Failed to create sigint event source: %m");
+                goto finish;
+        }
+
+        r = sd_event_set_watchdog(event, true);
+        if (r < 0) {
+                log_error_errno(r, "Failed to create watchdog event source: %m");
+                goto finish;
+        }
+
+        r = clock_state_update(&state, event);
+        if (r > 0) {
+                r = sd_event_loop(event);
+                if (0 > r)
+                        log_error_errno(r, "Failed in event loop: %m");
+                else if (state.adjtime_state == TIME_ERROR) {
+                        log_error("Event loop terminated without synchronizing");
+                        r = -1;
```

Comment:
we generally try to avoid making up error codes such as -1. Use something like `-EPROTO` or so